README
== Introduction ==
IBus-Indic-Table is the IM Engine framework for table-based input methods, and is
forked from ibus-table whose Project Owner & Developer is Yu Yuwei <acevery@gmail.com>,
with contributions from Caius "kaio" Chance <cchance@redhat.com>.
The aim of IBus-Indic-Table is to provide a table based boosted environment
for SCIM style tables. This project started out as an effort on idea and algorithm 
by Pravin Satpute <psatpute@redhat.com> which is described at 
https://fedoraproject.org/wiki/Features/IndicTypingBooster &
https://fedorahosted.org/indic-typing-booster/
Definitly, it is far from perfect, so if you have any wish or any
suggestion, please send me an email <nav007@gmail.com>, I would be very please to hear
from you :)

== Key Bindings ==
    Notice: Ctrl is "Control Key", Shift_L is the left "Shift Key",
            Shift_R is the right "Shift Key", Alt is the "Alt Key",
            Arrow_Left is the "Arrow Key" point to left, Arrow_Right is
            the "Arrow Key" point to right, similarly as  Arrow_Up and
            Arrow_Down.
            Cursor is the one in preedit string.
            [Num1~Num9] means one key in Num1, Num2, Num3, ... ,
                Num9

Space               commit inputted string.
Shift_L             when no key inputted: shift the input mode between
                        English and Table mode.
                    during input: commit the first candidate into
                        preedit string.

Arrow_Left          move cursor one character left in preedit string.
Arrow_Right         move cursor one character right in preedit string.
Backspace           remove the last input key or last character in
                        preeidt string.
Delete              remove one character after the cursor in preedit
                        string.
Ctrl + Arrow_Lelf   move the cursor to the front of preedit string.
Ctrl + Arrow_Right  move the cursor to the end of preedit string.
Ctrl + Backspace    delete from cursor to the front of preedit string.
Ctrl + Delete       delete from cursor to the end of preedit string.
Arrow_Down          mext candidate.
Arrow_Up            previous candidate.
Esc                 reset the input method.
Num1 ~ Num9         select the correspondent candidate and commit
                        inputted string
Ctrl + [Num1~Num9]  select the correspondent candidate and commit it
                        into preedit string
Alt + [Num1~Num9]   remove the correspondent user-defined phrase.
Page_Up/Page_Down   page up/down the lookup table of candidates.
-/+                 when not been used as valid input keys, the same as
                        Page_Up and Page_Down.
Ctrl + ,            open/close Single Character Mode, only show single
                        character not phrases.
Ctrl + .            shift between full/half punctuation.
Ctrl + /            shift between direct/normal commit mode.
Ctrl + ;            shift between different SubCharset in Chinese
Ctrl + '            show/hide speedmeter


ChangeLog
commit ec7148178b2c0c5c6e22ae6ee5b0e8d7f2c89a3c
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:51:03 2011 +0530

    oops forgot to add Pravin's name

commit d67f2c8d83f93309b78ca00e80e85d96c47f3d33
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:46:32 2011 +0530

    new ChangeLog generated

commit 4f21b33635f560933bc44fd551e961c67740b568
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:45:47 2011 +0530

    new uuid

commit e6ba1f93eaba4e98e4974bcd6970a4da33b65075
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:40:23 2011 +0530

    changes in README

commit 97f44869808a9b32a2d41888b46e40236942a85b
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:39:18 2011 +0530

    changes in README

commit dfea8e8a86c0947eeca19e17e567491683e82873
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:30:22 2011 +0530

    generate new ChangeLog

commit 46142b9a62af509255943637adf8b45e65a03120
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 16:05:03 2011 +0530

    changes in configure for bump in minor version

commit 4ec1222172940941073a2ed4a0f0f98ad783ef58
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 15:47:14 2011 +0530

    bump minor version

commit 724ddf00e2993cb85792c5eb4afb2debf2d5867a
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 15:43:16 2011 +0530

    changes in configure

commit 4866a7f46c0eedbfef5f3a9c9e461f08f5f36e0c
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 15:41:08 2011 +0530

    changes in configure.ac to exclude marathi tables

commit c7bea569754d6197c4d15b97e275dae6e13641fa
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 15:33:56 2011 +0530

    changes in README

commit aa4c44bce06258900610b9e4205a940334bebf20
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 8 15:20:39 2011 +0530

    remove pinyin specific data

commit 21cd305cb597f601c0ef0fe3aafb2503547a5bb4
Author: Naveen Kumar <nav007@gmail.com>
Date:   Fri Mar 4 15:32:59 2011 +0530

    changes in configure

commit 394a2363b34ea95d548193e5657c6d84a15b0df1
Author: Naveen Kumar <nav007@gmail.com>
Date:   Fri Mar 4 15:27:15 2011 +0530

    changes in configure.ac and remove execute permissions from all am,in & other nonexecutable files

commit b40ad23bfb1b6b1dbc52310072668af428de3d5a
Author: Naveen Kumar <nav007@gmail.com>
Date:   Wed Mar 2 19:30:05 2011 +0530

    remove python shebang from tabcreatedb

commit f40ce71c2b583600ab6cfc9dbb7754c50b370068
Author: Naveen Kumar <nav007@gmail.com>
Date:   Wed Mar 2 16:53:47 2011 +0530

    changes in Makefile.am Makefile.in removing latex.svg and compose.svg

commit 25b8907a07db97f026871a791c389d87d39d97c4
Author: Naveen Kumar <nav007@gmail.com>
Date:   Wed Mar 2 16:12:41 2011 +0530

    remove compose.svg latex.svg

commit c5b9ad2b86e4fb20a2e7f19d44c2fbeaccfcabf9
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 1 15:58:27 2011 +0530

    remove indic-typing-booster files and others

commit b8012353dc025111827ad5ace0eb67744ed4e573
Author: Naveen Kumar <nav007@gmail.com>
Date:   Tue Mar 1 15:54:39 2011 +0530

    commiting autotools changes & remove indic-typing-booster files and others

commit 85c58cb45faa283af6f9a8fe441caff7a14d45aa
Author: Naveen Kumar <nav007@gmail.com>
Date:   Thu Feb 24 16:24:55 2011 +0530

    oops forgot to modify confiure for indic typing booster

commit ce835361b9c3a200ba9dbc50fa06256b01ea201f
Author: Naveen Kumar <nav007@gmail.com>
Date:   Thu Feb 24 16:21:48 2011 +0530

    modified autotools files for including indic typing booster

commit 93084f5025c01fbd45827ed34377328de3bf2c6e
Author: Naveen Kumar <nav007@gmail.com>
Date:   Thu Feb 24 16:15:15 2011 +0530

    added marathi-inscript.txt marathi-phonetic.txt from indic typing booster

commit 5761a5c33c33fd10f4b7b54dedb8188be8001150
Author: Naveen Kumar <nav007@gmail.com>
Date:   Wed Feb 23 19:25:36 2011 +0530

    pushing changes made to ibus-table for to create ibus-indic-table

commit d425e611e4bc8fe7a569a9d6aa82b7c5959a2d60
Author: Naveen Kumar <nav007@gmail.com>
Date:   Wed Feb 23 19:23:05 2011 +0530

    first commit


engine/main.py
# vim:et sts=4 sw=4
#
# ibus-indic-table - The Tables engine for IBus
#
# Copyright (c) 2011-2012 Naveen Kumar <nav007@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#

import os
import sys
import optparse
import ibus
import gobject
import re
patt = re.compile (r'<\?.*\?>\n')
from signal import signal, SIGTERM, SIGINT

import factory
import tabsqlitedb


try:
    db_dir = os.path.join (os.getenv('IBUS_INDIC_TABLE_LOCATION'),'indic-tables')
    icon_dir = os.path.join (os.getenv('IBUS_INDIC_TABLE_LOCATION'),'icons')
except:
    db_dir = "/usr/share/ibus-indic-table/indic-tables"
    icon_dir = "/usr/share/ibus-indic-table/icons"


opt = optparse.OptionParser()

opt.set_usage ('%prog --table a_table.db')
opt.add_option('--table', '-t',
        action = 'store',type = 'string',dest = 'db',default = '',
        help = 'Set the IME table file, default: %default')
opt.add_option('--daemon','-d',
        action = 'store_true',dest = 'daemon',default=False,
        help = 'Run as daemon, default: %default')
opt.add_option('--ibus', '-i',
        action = 'store_true',dest = 'ibus',default = False,
        help = 'Set the IME icon file, default: %default')
opt.add_option('--xml', '-x',
        action = 'store_true',dest = 'xml',default = False,
        help = 'output the engines xml part, default: %default')

opt.add_option('--no-debug', '-n',
        action = 'store_false',dest = 'debug',default = True,
        help = 'redirect stdout and stderr to ~/.ibus/indic-tables/debug.log, default: %default')

(options, args) = opt.parse_args()
#if not options.db:
#    opt.error('no db found!')

if (not options.xml) and options.debug:
    if not os.access ( os.path.expanduser('~/.ibus/indic-tables'), os.F_OK):
        os.system ('mkdir -p ~/.ibus/indic-tables')
    logfile = os.path.expanduser('~/.ibus/indic-tables/debug.log')
    sys.stdout = open (logfile,'a',0)
    sys.stderr = open (logfile,'a',0)
    from time import strftime
    print '--- ', strftime('%Y-%m-%d: %H:%M:%S'), ' ---'



class IMApp:
    def __init__(self, dbfile, exec_by_ibus):
        self.__mainloop = gobject.MainLoop()
        self.__bus = ibus.Bus()
        self.__bus.connect("disconnected", self.__bus_destroy_cb)
        self.__factory = factory.EngineFactory(self.__bus, dbfile)
        self.destroied = False
        if exec_by_ibus:
            self.__bus.request_name("org.freedesktop.IBus.IndicTable", 0)
        else:
            self.__component = ibus.Component("org.freedesktop.IBus.IndicTable",
                                              "Table Component",
                                              "0.1.0",
                                              "GPL",
                                              "Naveen Kumar <nav007@gmail.com>")
            # now we get IME info from self.__factory.db
            name = self.__factory.db.get_ime_property ("name")
            longname = name
            description = self.__factory.db.get_ime_property ("description")
            language = self.__factory.db.get_ime_property ("languages")
            license = self.__factory.db.get_ime_property ("credit")
            author = self.__factory.db.get_ime_property ("author")
            icon = self.__factory.db.get_ime_property ("icon")
            if icon:
                icon = os.path.join (icon_dir, icon)
                if not os.access( icon, os.F_OK):
                    icon = ''
            layout = self.__factory.db.get_ime_property ("layout")
            
            self.__component.add_engine(name,
                                        longname,
                                        description,
                                        language,
                                        license,
                                        author,
                                        icon,
                                        layout)
            self.__bus.register_component(self.__component)


    def run(self):
        self.__mainloop.run()
        self.__bus_destroy_cb()

    def quit(self):
        self.__bus_destroy_cb()

    def __bus_destroy_cb(self, bus=None):
        if self.destroied:
            return
        print "finalizing:)"
        self.__factory.do_destroy()
        self.destroied = True
        self.__mainloop.quit()

def cleanup (ima_ins):
    ima_ins.quit()
    sys.exit()

def indent(elem, level=0):
    '''Use to format xml Element pretty :)'''
    i = "\n" + level*"    "
    if len(elem):
        if not elem.text or not elem.text.strip():
            elem.text = i + "    "
        for e in elem:
            indent(e, level+1)
            if not e.tail or not e.tail.strip():
                e.tail = i + "    "
        if not e.tail or not e.tail.strip():
            e.tail = i
    else:
        if level and (not elem.tail or not elem.tail.strip()):
            elem.tail = i

def main():
    if options.xml:
        from locale import getdefaultlocale
        from xml.etree.ElementTree import Element, SubElement, tostring
        # we will output the engines xml and return.
        # 1. we find all dbs in db_dir and extract the infos into
        #    Elements
        dbs = os.listdir(db_dir)
        dbs = filter (lambda x: x.endswith('.db'), dbs)
        #if not dbs:
        #    return

        egs = Element('engines')
        for _db in dbs:
            _sq_db = tabsqlitedb.tabsqlitedb (os.path.join (db_dir, _db))
            _engine = SubElement (egs,'engine')
            
            _name = SubElement (_engine, 'name')
            _name.text = _db.replace ('.db','')
            
            _longname = SubElement (_engine, 'longname')
            _longname.text = ''
            try:
                _locale = getdefaultlocale()[0].lower()
                _longname.text = _sq_db.get_ime_property ( \
                    '.'.join(['name',_locale]) )
            except:
                pass
            if not _longname.text:
                _longname.text = _name.text
            
            _language = SubElement (_engine, 'language')
            _langs = _sq_db.get_ime_property ('languages')
            if _langs:
                _langs = _langs.split (',')
                if len (_langs) == 1:
                    _language.text = _langs[0].strip()
                else:
                    # we ignore the place
                    _language.text = _langs[0].strip().split('_')[0]

            _license = SubElement (_engine, 'license')
            _license.text = _sq_db.get_ime_property ('license')

            _author = SubElement (_engine, 'author')
            _author.text  = _sq_db.get_ime_property ('author')

            _icon = SubElement (_engine, 'icon')
            _icon_basename = _sq_db.get_ime_property ('icon')
            if _icon_basename:
                _icon.text = os.path.join (icon_dir, _icon_basename)
            
            _layout = SubElement (_engine, 'layout')
            _layout.text = _sq_db.get_ime_property ('layout')

            _desc = SubElement (_engine, 'description')
            _desc.text = _sq_db.get_ime_property ('description')

        # now format the xmlout pretty
        indent (egs)
        egsout = tostring (egs, encoding='utf8')
        egsout = patt.sub ('',egsout)
        print egsout
        
        return 0

    if options.daemon :
        if os.fork():
                sys.exit()
    if options.db:
        if os.access( options.db, os.F_OK):
            db = options.db
        else:
            db = '%s%s%s' % (db_dir,os.path.sep, os.path.basename(options.db) )
    else:
        db=""
    ima=IMApp(db, options.ibus)
    signal (SIGTERM, lambda signum, stack_frame: cleanup(ima))
    signal (SIGINT, lambda signum, stack_frame: cleanup(ima))
    try:
        ima.run()
    except KeyboardInterrupt:
        ima.quit()

if __name__ == "__main__":
    main()



engine/factory.py
# -*- coding: utf-8 -*-
# vim:et sw=4 sts=4 sw=4
#
# ibus-indic-table - The Tables engine for IBus
#
# Copyright (c) 2011-2012 Naveen Kumar <nav007@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# $Id: $
#

import ibus
from ibus.exception import *
import indic_table
import tabsqlitedb
import os
import dbus
from re import compile as re_compile

path_patt = re_compile(r'[^a-zA-Z0-9_/]')

from gettext import dgettext
_  = lambda a : dgettext ("ibus-indic-table", a)
N_ = lambda a : a

engine_base_path = "/com/redhat/IBus/engines/table/%s/engine/"


class EngineFactory (ibus.EngineFactoryBase):
    """Table IM Engine Factory"""
    def __init__ (self, bus, db="", icon=""):
        # here the db should be the abs path to sql db
        # this is the backend sql db we need for our IME
        # we need get lots of IME property from this db :)
        #self.db = tabsqlitedb.tabsqlitedb( name = database )
        
        if db:
            self.dbusname = os.path.basename(db).replace('.db','')
            udb = os.path.basename(db).replace('.db','-user.db') 
            self.db = tabsqlitedb.tabsqlitedb( name = db,user_db = udb )
            self.db.db.commit()
            self.dbdict = {self.dbusname:self.db}
        else:
            self.db = None
            self.dbdict = {}

        
        # init factory
        self.bus = bus
        super(EngineFactory,self).__init__ (bus)
        self.engine_id=0
        try:
            bus = dbus.Bus()
            user = os.path.basename( os.path.expanduser('~') )
            self._sm_bus = bus.get_object ("org.ibus.table.SpeedMeter.%s"\
                    % user, "/org/ibus/table/SpeedMeter")
            self._sm =  dbus.Interface(self._sm_bus,\
                    "org.ibus.table.SpeedMeter") 
        except:
            self._sm = None
    
    def create_engine(self, engine_name):
        # because we need db to be past to Engine
        # the type (engine_name) == dbus.String
        name = engine_name.encode ('utf8')
        self.engine_path = engine_base_path % path_patt.sub ('_', name)
        try:
            if not self.db:
                # first check self.dbdict
                if not name in self.dbdict:
                    try:
                        db_dir = os.path.join (os.getenv('IBUS_INDIC_TABLE_LOCATION'),'indic-tables')
                    except:
                        db_dir = "/usr/share/ibus-indic-table/indic-tables"
                    db = os.path.join (db_dir,name+'.db')
                    udb = name+'-user.db'
                    _sq_db = tabsqlitedb.tabsqlitedb( name = db,user_db = udb )
                    _sq_db.db.commit()
                    self.dbdict[name] = _sq_db
            else:
                name = self.dbusname

            engine = indic_table.tabengine(self.bus, self.engine_path \
                    + str(self.engine_id), self.dbdict[name])
            self.engine_id += 1
            #return engine.get_dbus_object()
            return engine
        except:
            print "fail to create engine %s" % engine_name
            import traceback
            traceback.print_exc ()
            raise IBusException("Can not create engine %s" % engine_name)

    def do_destroy (self):
        '''Destructor, which finish some task for IME'''
        # 
        ## we need to sync the temp userdb in memory to the user_db on disk
        for _db in self.dbdict:
            self.dbdict[_db].sync_usrdb ()
        ##print "Have synced user db\n"
        try:
            self._sm.Quit()
        except:
            pass
        super(EngineFactory,self).do_destroy()




engine/tabdict.py
# -*- coding: utf-8 -*-
# vim:et sts=4 sw=4
#
# ibus-indic-table - The Tables engine for IBus
#
# Copyright (c) 2011-2012 Naveen Kumar <nav007@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# $Id: $
#
import sys
reload (sys)
sys.setdefaultencoding('utf-8')

tab_dict = {
    '0':0,
    'a':1, 'b':2, 'c':3, 'd':4, 'e':5, 
    'f':6, 'g':7, 'h':8, 'i':9, 'j':10,
    'k':11, 'l':12, 'm':13, 'n':14, 'o':15,
    'p':16, 'q':17, 'r':18, 's':19, 't':20,
    'u':21, 'v':22, 'w':23, 'x':24, 'y':25,
    'z':26, "'":27, ';':28, '`':29, '~':30, 
    '!':31, '@':32, '#':33, '$':34, '%':35,
    '^':36, '&':37, '*':38, '(':39, ')':40,
    '-':41, '_':42, '=':43, '+':44, '[':45,
    ']':46, '{':47, '}':48, '|':49, '/':50,
    ':':51, '"':52,  '<':53, '>':54, ',':55,
    '.':56, '?':57, '\\':58, 'A':59, 'B':60,
    'C':61, 'D':62, 'E':63, 'F':64, 'G':65,
    'H':66, 'I':67, 'J':68, 'K':69, 'L':70,
    'M':71, 'N':72, 'O':73, 'P':74, 'Q':75,
    'R':76, 'S':77, 'T':78, 'U':79, 'V':80,
    'W':81, 'X':82, 'Y':83, 'Z':84, '0':85,
    '1':86, '2':87, '3':88, '4':89, '5':90,
    '6':91, '7':92, '8':93, '9':94
    }

tab_key_list = tab_dict.keys()

id_tab_dict = {}
for key,id in tab_dict.items():
    id_tab_dict[id] = key

class tab_key(object):
    '''The class store'''
    def __init__(self, xm_key):
        self._key = xm_key
        try:
            self._key_id = tab_dict[xm_key]
        except KeyError, e:
            # give a false value
            self._key_id = -1
            error_m = u'%s is not in tab_dict' % xm_key
            print error_m.encode('utf8')
            import traceback
            traceback.print_exc ()
    
    def get_key_id(self):
        return self._key_id

    def get_key(self):
        return self._key

    def __str__(self):
        return self._key

    def __int__(self):
        return self._key_id

def parse ( inputstr ):
    
    ids_input = []
    try:
        ids_input = map (tab_key,inputstr)
    except:
        pass
    return ids_input[:]

def deparse (id):
    '''deparse the id code of tables, id could be int or int in string form'''
    if id:
        id = int(id)
        if id in id_tab_dict:
            return id_tab_dict[id]
    else:
        return ''

chinese_nc_chars="“”‘’《》〈〉〔〕「」『』【】〖〗（）［］｛｝"\
    "．。，、；：？！…—·ˉˇ¨々～‖∶＂＇｀｜"\
    "⒈⒉⒊⒋⒌⒍⒎⒏⒐⒑⒒⒓⒔⒕⒖⒗⒘⒙⒚⒛"\
    "АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯЁ"\
    "ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫ"\
    "⒈⒉⒊⒋⒌⒍⒎⒏⒐⒑⒒⒓⒔⒕⒖⒗⒘⒙⒚⒛"\
    "㎎㎏㎜㎝㎞㎡㏄㏎㏑㏒㏕"\
    "ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩ"\
    "⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇"\
    "€＄￠￡￥"\
    "¤→↑←↓↖↗↘↙"\
    "ァアィイゥウェエォオカガキギクグケゲコゴサザシジ"\
    "スズセゼソゾタダチヂッツヅテデトドナニヌネノハバパ"\
    "ヒビピフブプヘベペホボポマミムメモャヤュユョヨラ"\
    "リルレロヮワヰヱヲンヴヵヶーヽヾ"\
    "ぁあぃいぅうぇえぉおかがきぎぱくぐけげこごさざしじ"\
    "すずせぜそぞただちぢっつづてでとどなにぬねのはば"\
    "ひびぴふぶぷへべぺほぼぽまみむめもゃやゅゆょよらり"\
    "るれろゎわゐゑをん゛゜ゝゞ"\
    "勹灬冫艹屮辶刂匚阝廾丨虍彐卩钅冂冖宀疒肀丿攵凵犭"\
    "亻彡饣礻扌氵纟亠囗忄讠衤廴尢夂丶"\
    "āáǎàōóǒòêēéěèīíǐìǖǘǚǜüūúǔù"\
    "＋－＜＝＞±×÷∈∏∑∕√∝∞∟∠∣∥∧∨∩∪∫∮"\
    "∴∵∶∷∽≈≌≒≠≡≤≥≦≧≮≯⊕⊙⊥⊿℃°‰"\
    "♂♀§№☆★○●◎◇◆□■△▲※〓＃＆＠＼＾＿￣"\
    "абвгдежзийклмнопрстуфхцчшщъыьэюяё"\
    "ⅰⅱⅲⅳⅴⅵⅶⅷⅸⅹβγδεζηαικλμνξοπρστυφθψω"\
    "①②③④⑤⑥⑦⑧⑨⑩①②③④⑤⑥⑦⑧⑨⑩"\
    "㈠㈡㈢㈣㈤㈥㈦㈧㈨㈩㈠㈡㈢㈣㈤㈥㈦㈧㈨㈩"\
    "ㄅㄆㄇㄈㄉㄊㄋㄌㄍㄎㄏㄐㄑㄒㄓㄔㄕㄖㄗㄘㄙㄧㄨㄩ"\
    "ㄚㄛㄜㄝㄞㄟㄠㄡㄢㄣㄤㄥㄦ"
chinese_nocheck_chars = chinese_nc_chars.decode('utf8')
